VnodeCacheDesign.md

```diff
@@ -0,0 +1,101 @@
+# Vnode Cache Design
+
+## Goals
+
+- Reduce the time required to determine if a file is inside a virtualization root
+- Reduce the time required to find a file's virtualization root
+- Eliminate the use of file flags to determine if a file is inside of a root
+
+## Design 
+
+### Cached Data
+
+PrjFSKext will cache the following information for each `vnode_t` pointer:
+
+- The `vnode_t` itself
+- The vnode's `uint32_t` vid (vnode generation number)
+- A `uint16_t` virtualization\cache generation number (hereby referred to as `vrgid`).  If a cache entry's `vrgid` does not match its roots it means that all entries for that root have been invalidated and the `VirtualizationRootHandle` in the cache entry is no longer valid.  There will also be a global `vrgid`  for all vnodes that are not inside of any virtualization roots.
+- `VirtualizationRootHandle` for the vnode (or `RootHandle_None` if the vnode is not in a root).  **NOTE** This assumes that `VirtualizationRootHandle`  are not recycled as VFS4G instances are unmounted\remounted.
+
+### Hash Table
+
+The data for each vnode will be stored in a fixed-size hash table with the `vnode_t` pointer serving as the key.
+
+#### Collision Handling
+
+Index collisions will be handled using open addressing.  The exact technique (e.g. linear probing, double hashing, etc.) will be determined once we have sample sets of `vnode_t` pointers and can evaluate the quality of our hash function.
+
+#### Size
+
+The hash table will be a fixed size that is some percent larger than the value of `desiredvnodes` (exact percentage is still TBD).
+
```

Comment:
The Collision Handing, Table Size, and Hash Key are pretty key pieces.  Are there similar drivers out there we can follow for guidance?

Reply:
Not that I'm aware of.  @nickgra, @sanoursa, and I recently met with two other teams building similar kexts and neither are caching all of the vnodes.